controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp

```diff
@@ -46,14 +46,142 @@ dynamic_channel_selection_task::dynamic_channel_selection_task(db &database_,
+        }
+        break;
+    }
+    case eState::TRIGGER_SCAN: {
+        LOG(TRACE) << "TRIGGER_SCAN, mac=" << m_radio_mac << "scan_type is "
+                   << ((m_is_single_scan) ? "single-scan" : "continuous-scan");
+
+        // When a scan is requested send the scan parameters Channel pool & Dwell time
+        // Before sending the request set the scan_in_progress flag to true
+        // So another scan would not trigger on the same radio simultaneously
+
+        database.set_channel_scan_in_progress(m_radio_mac, true, m_is_single_scan);
+
+        auto request = message_com::create_vs_message<
+            beerocks_message::cACTION_CONTROL_CHANNEL_SCAN_TRIGGER_SCAN_REQUEST>(cmdu_tx);
+        if (request == nullptr) {
+            LOG(ERROR) << "Failed building message cACTION_CONTROL_CHANNEL_SCAN_TRIGGER_SCAN_REQUEST !";
+            m_last_scan_error_code = beerocks::CHANNEL_SCAN_INTERNAL_FAILURE;
+            FSM_MOVE_STATE(ABORT_SCAN);
+            break;
+        }
+
+        int32_t dwell_time_msec = database.get_channel_scan_dwell_time_msec(m_radio_mac, m_is_single_scan);
+        if (dwell_time_msec < 0) {
+            LOG(ERROR) << "invalid dwell_time < 0";
+            m_last_scan_error_code = beerocks::CHANNEL_SCAN_INVALID_PARAMS;
+            FSM_MOVE_STATE(ABORT_SCAN);
+            break;
+        }
+
+        //get current channel pool from DB
+        auto &curr_channel_pool = database.get_channel_scan_pool(m_radio_mac, m_is_single_scan);
+        if (curr_channel_pool.empty()) {
+            LOG(ERROR) << "empty channel pool is not supported. please set channel pool for mac="
+                       << m_radio_mac;
+            m_last_scan_error_code = beerocks::CHANNEL_SCAN_INVALID_PARAMS;
+            FSM_MOVE_STATE(ABORT_SCAN);
+            break;
+        }
+
+        if (curr_channel_pool.size() > beerocks::message::SUPPORTED_CHANNELS_LENGTH) {
+            LOG(ERROR) << "channel_pool is too big";
+            m_last_scan_error_code = beerocks::CHANNEL_SCAN_POOL_TOO_BIG;
+            FSM_MOVE_STATE(ABORT_SCAN);
+            break;
+        }
+
+        request->scan_params().radio_mac         = net::network_utils::mac_from_string(m_radio_mac);
+        request->scan_params().dwell_time_ms     = dwell_time_msec;
+        request->scan_params().channel_pool_size = curr_channel_pool.size();
+        std::copy(curr_channel_pool.begin(), curr_channel_pool.end(),
+                  request->scan_params().channel_pool);
+
+        // get the parent node to send the CMDU to the agent
+        auto ire = database.get_node_parent_ire(m_radio_mac);
+        son_actions::send_cmdu_to_agent(ire, cmdu_tx, database, m_radio_mac);
+
+        set_events_timeout(SCAN_TRIGGERED_WAIT_TIME_MSEC);
+        dcs_wait_for_event(eEvent::SCAN_TRIGGERED);
+
+        FSM_MOVE_STATE(WAIT_FOR_SCAN_TRIGGERED);
+        break;
+    }
+    case eState::WAIT_FOR_SCAN_TRIGGERED: {
+        TASK_LOG(ERROR) << "This should not happen!";
```

Comment:
OK, why this shouldn't happen though? should be a comment explaining this. I really don't understand, you move the FSM to this state so the next work iteration you will get here..

Reply:
in ::work()
we TRIGGER_SCAN and got to WAIT_FOR_SCAN_TRIGGERED state and set the timout.
in  handle_event()
if  we dont recieve SCAN_TRIGGERED event from radio thats why we need the WAIT_FOR_SCAN_TRIGGERED  timout.

same of:
    case eState::WAIT_FOR_RESULTS_READY:
    case eState::WAIT_FOR_RESULTS_DUMP:

Reply:
It's less a "This should not happen" and more on the lines of "If this happens it means something somewhere went wrong"

The WAIT_FOR_ states are for timeout handling